cli/azd/pkg/environment/environment.go

```diff
@@ -36,14 +37,78 @@ const ContainerRegistryEndpointEnvVarName = "AZURE_CONTAINER_REGISTRY_ENDPOINT"
 const ResourceGroupEnvVarName = "AZURE_RESOURCE_GROUP"
 
 type Environment struct {
-	// Values is a map of setting names to values.
-	Values map[string]string
+	// values is a map of setting names to values.
+	// all keys are converted to upper case before it is set.
+	values map[string]string
 	// File is a path to the file that backs this environment. If empty, the Environment
 	// will not be persisted when `Save` is called. This allows the zero value to be used
 	// for testing.
 	File string
 }
 
+// DeleteVariable removes an entry from azd environment which key matches the variable name.
+func (e *Environment) DeleteVariable(variableName string) {
+	delete(e.values, strings.ToUpper(variableName))
+}
+
+// SetVariable update or create an entry to the azd environment.
+// key is converted to upper case.
+func (e *Environment) SetVariable(key, value string) {
+	e.values[strings.ToUpper(key)] = value
+}
+
+// GetValue get the value of an entry with the requested key.
+// return empty string and false if the key is not found.
+// Use ValueOf if you don't need to check if value is found
+func (e *Environment) GetValue(key string) (string, bool) {
+	value, found := e.values[strings.ToUpper(key)]
+	return value, found
+}
+
+// GetValue get the value of an entry with the requested key.
+// return empty string and false if the key is not found.
+// Use GetValue to return if the value is found as well
+func (e *Environment) ValueOf(key string) string {
+	return e.values[strings.ToUpper(key)]
+}
+
+// HasValue return true if the key is found and the value is not empty string.
+func (e *Environment) HasValue(key string) bool {
+	value, found := e.values[strings.ToUpper(key)]
+	return found && value != ""
+}
+
+// Init make sure that values has an empty map.
+func (e *Environment) Init() {
```

Comment:
Should we apply this as an invariant at construction time? This might be error-prone unless we perform nil-checks everywhere

Reply:
Adding a constructor `NewEnvironment` to handle this initialization would be good, but it comes with a little price.
We need to make the `Environment` an interface (and not an struct any more) and hide the current implementation behind the constructor (basically the same approach we do for `AzCli`).

The side extra benefit is that it would help with testing. As we wiil be able to create mocked environment.

I will implement the change